lib/grpc/server/interceptors/cors.ex

```diff
@@ -0,0 +1,97 @@
+defmodule GRPC.Server.Interceptors.CORS do
+  @moduledoc """
+  Sends CORS headers when the client is using RPC via Web transcoding or gRPC-web.
+
+  ## Options
+
+    * `allow_origin` - Required. A string containing the allowed origin, or a remote function (e.g. `&MyApp.MyModule.function/2)`) which takes a `req` and a `stream` and returns a string.
+    *`allow_headers` - A string containing the allowed headers, or a remote function (e.g. `&MyApp.MyModule.function/2)`) which takes a `req` and a `stream` and returns a string. If not defined, the value of the `"access-control-request-headers"` request header from the client will be used in the response.
+
+  ## Usage
+
+      defmodule Your.Endpoint do
+        use GRPC.Endpoint
+
+        intercept GRPC.Server.Interceptors.CORS
+      end
+
+      defmodule Your.Endpoint do
+        use GRPC.Endpoint
+
+        intercept GRPC.Server.Interceptors.CORS, allow_origin: "some.origin"
+      end
+
+
+      defmodule Your.Endpoint do
+        use GRPC.Endpoint
+
+        def allow_origin(req, stream), do: "calculated.origin"
+        intercept GRPC.Server.Interceptors.CORS, allow: &Your.Endpoint.allow_origin/2
+      end
+  """
+
+  @behaviour GRPC.Server.Interceptor
+  @impl true
+  def init(opts \\ []) do
+    # the funky first clause matches a 2-arity remote is_function
+    # note that this function is run in the context of a macro, which brings some limitations with it
+    allow_origin =
+      case Keyword.get(opts, :allow_origin) do
+        {:&, [], [{:/, [], [_signature, 2]}]} = fun -> fun
+        static when is_binary(static) -> static
```

Comment:
You can add a explicit catch all here raising and ArgumentError for better UX.

This matching mode is also already enough to guard against runtime functions as I'd mentioned below.

Reply:
> You can add a explicit catch all here raising and ArgumentError for better UX.

see b51feda

> This matching mode is also already enough to guard against runtime functions as I'd mentioned below.

When called outside of a macro, yes. When called in a macro, it fails at compilation before it even gets there.